src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -3340,6 +3332,73 @@ private RubyNode addNewlineIfNeeded(ParseNode jrubyNode, RubyNode node) {
         return node;
     }
 
+    private interface ValueFromNode {
+
+        RubyNode prepareAndThen(SourceIndexLength sourceSection, RubyNode subsequent);
+
+        ParseNode get(SourceIndexLength sourceSection);
+
+    }
+
+    private class ValueFromEffectNode implements ValueFromNode {
```

Comment:
We should avoid non-static inner classes as much as possible, they tend to be harder to read, and have unintuitive GC semantics.
In this case, it's clearer to pass the `environment` explicitly.

Reply:
Can do.

Reply:
Fixed.